File: sdk/python/packages/flet-core/src/flet_core/semantics_service.py
from enum import Enum
from typing import Any, Optional

from flet_core.control import Control
from flet_core.ref import Ref


class Assertiveness(Enum):
    POLITE = "polite"
    ASSERTIVE = "assertive"


class SemanticsService(Control):
    def __init__(
        self,
        ref: Optional[Ref] = None,
        data: Any = None,
    ):
        Control.__init__(
            self,
            ref=ref,
            data=data,
        )

    def _get_control_name(self):
        return "semanticsservice"

    def announce_message(
        self,
        message: str,
        rtl: bool = False,
        assertiveness: Assertiveness = Assertiveness.POLITE,
    ):
        self.invoke_method(
            "announce_message",
            arguments={
                "message": message,
                "rtl": str(rtl),
                "assertiveness": assertiveness.value
                if isinstance(assertiveness, Assertiveness)
                else str(assertiveness),
            },
        )

    def announce_tooltip(self, message: str):
        self.invoke_method(
            "announce_tooltip",
            arguments={
                "message": message,
            },
        )

File: packages/flet/lib/src/services/semantics_service.dart
import 'package:flutter/material.dart';
import 'package:flutter/semantics.dart';

import '../flet_service.dart';
import '../utils/misc.dart';

class SemanticsServiceControl extends FletService {
  SemanticsServiceControl({required super.control});

  @override
  void init() {
    super.init();
    debugPrint("SemanticsService(${control.id}).init: ${control.properties}");
    control.addInvokeMethodListener(_invokeMethod);
  }

  @override
  void dispose() {
    debugPrint("SemanticsService(${control.id}).dispose()");
    control.removeInvokeMethodListener(_invokeMethod);
    super.dispose();
  }

  Future<dynamic> _invokeMethod(String name, dynamic args) async {
    debugPrint("SemanticsService.$name($args)");
    switch (name) {
      case "announce_message":
        var message = args["message"].toString();
        final view = WidgetsBinding.instance.platformDispatcher.implicitView;
        if (view == null) {
          return;
        }
        return SemanticsService.sendAnnouncement(
            view, message, args["rtl"] ? TextDirection.rtl : TextDirection.ltr,
            assertiveness: control.getAssertiveness(
                args["assertiveness"], Assertiveness.polite)!);
      case "announce_tooltip":
        var message = args["message"].toString();
        return SemanticsService.tooltip(message);
      case "get_accessibility_features":
        var features = SemanticsBinding.instance.accessibilityFeatures;
        return {
          "accessible_navigation": features.accessibleNavigation,
          "bold_text": features.boldText,
          "disable_animations": features.disableAnimations,
          "high_contrast": features.highContrast,
          "invert_colors": features.invertColors,
          "reduce_motion": features.reduceMotion,
          "on_off_switch_labels": features.onOffSwitchLabels,
          "supports_announcements": features.supportsAnnounce,
        };
      default:
        throw Exception("Unknown SemanticsService method: $name");
    }
  }
}
